File: backend/app/routes/portfolios.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import or_
from sqlalchemy.orm import Session

from app.auth.deps import get_approved_user
from app.database import get_db
from app.models import AuditLog, Portfolio, User, UserRole
from app.schemas import (
    PortfolioCreate,
    PortfolioListItem,
    PortfolioListResponse,
    PortfolioOut,
    PortfolioUpdate,
)
from app.services import quota


router = APIRouter(prefix="/portfolios", tags=["portfolios"])


def _to_out(p: Portfolio, owner_name: str) -> PortfolioOut:
    return PortfolioOut(
        id=p.id,
        user_id=p.user_id,
        owner_name=owner_name,
        name=p.name,
        portfolio_type=p.portfolio_type,
        risk_tolerance=p.risk_tolerance,
        initial_capital=p.initial_capital,
        target_return=p.target_return,
        target_risk=p.target_risk,
        risk_free_rate=p.risk_free_rate,
        history_years=p.history_years,
        min_history_years=p.min_history_years,
        cov_method=p.cov_method,
        weights=p.weights or [],
        universe_size=p.universe_size,
        sparsified=p.sparsified,
        expected_return_annual=p.expected_return_annual,
        volatility_annual=p.volatility_annual,
        sharpe_ratio=p.sharpe_ratio,
        sortino_ratio=p.sortino_ratio,
        var_95_annual=p.var_95_annual,
        cvar_95_annual=p.cvar_95_annual,
        max_drawdown_estimate=p.max_drawdown_estimate,
        monte_carlo=p.monte_carlo,
        efficient_frontier=p.efficient_frontier,
        correlation_matrix=p.correlation_matrix,
        benchmark_comparison=p.benchmark_comparison,
        is_public=p.is_public,
        notes=p.notes,
        created_at=p.created_at,
    )


@router.post("", response_model=PortfolioOut)
def create_portfolio(
    payload: PortfolioCreate,
    user: User = Depends(get_approved_user),
    db: Session = Depends(get_db),
):
    allowed, reason, _ = quota.check_can_generate(db, user)
    if not allowed:
        raise HTTPException(status_code=429, detail=reason)

    res = payload.optimize_result
    req = payload.optimize_request

    p = Portfolio(
        user_id=user.id,
        name=payload.name,
        portfolio_type=res.portfolio_type,
        risk_tolerance=req.risk_tolerance,
        initial_capital=res.initial_capital,
        target_return=req.target_return,
        target_risk=req.target_risk,
        risk_free_rate=res.risk_free_rate,
        history_years=res.history_years,
        min_history_years=res.min_history_years,
        cov_method=res.cov_method,
        weights=[w.model_dump() for w in res.weights],
        universe_size=res.universe_size,
        sparsified=res.sparsified,
        expected_return_annual=res.expected_return_annual,
        volatility_annual=res.volatility_annual,
        sharpe_ratio=res.sharpe_ratio,
        sortino_ratio=res.sortino_ratio,
        var_95_annual=res.var_95_annual,
        cvar_95_annual=res.cvar_95_annual,
        max_drawdown_estimate=res.max_drawdown_estimate,
        monte_carlo=res.monte_carlo.model_dump() if hasattr(res.monte_carlo, "model_dump") else res.monte_carlo,
        efficient_frontier=res.efficient_frontier,
        correlation_matrix=res.correlation_matrix,
        benchmark_comparison=res.benchmark_comparison,
        is_public=payload.is_public,
        notes=payload.notes,
    )
    db.add(p)
    db.commit()
    db.refresh(p)

    quota.record_generation(db, user.id, p.id)
    db.add(AuditLog(user_id=user.id, action="save_portfolio", detail=f"#{p.id} {p.name}"))
    db.commit()

    return _to_out(p, user.name)


@router.get("", response_model=PortfolioListResponse)
def list_portfolios(
    user: User = Depends(get_approved_user),
    db: Session = Depends(get_db),
    show_public: bool = True,
):
    q = db.query(Portfolio, User).join(User, Portfolio.user_id == User.id)
    if user.role == UserRole.ADMIN.value:
        rows = q.order_by(Portfolio.created_at.desc()).all()
    else:
        if show_public:
            rows = q.filter(or_(Portfolio.user_id == user.id, Portfolio.is_public.is_(True))) \
                    .order_by(Portfolio.created_at.desc()).all()
        else:
            rows = q.filter(Portfolio.user_id == user.id) \
                    .order_by(Portfolio.created_at.desc()).all()

    items = []
    for p, u in rows:
        items.append(PortfolioListItem(
            id=p.id, user_id=p.user_id, owner_name=u.name, name=p.name,
            portfolio_type=p.portfolio_type, initial_capital=p.initial_capital,
            expected_return_annual=p.expected_return_annual,
            volatility_annual=p.volatility_annual, sharpe_ratio=p.sharpe_ratio,
            is_public=p.is_public, is_mine=(p.user_id == user.id),
            created_at=p.created_at,
        ))
    return PortfolioListResponse(portfolios=items, total=len(items))


@router.get("/{pid}", response_model=PortfolioOut)
def get_portfolio(pid: int, user: User = Depends(get_approved_user), db: Session = Depends(get_db)):
    p = db.get(Portfolio, pid)
    if not p:
        raise HTTPException(status_code=404, detail="Not found")
    if p.user_id != user.id and not p.is_public and user.role != UserRole.ADMIN.value:
        raise HTTPException(status_code=403, detail="Forbidden")
    owner = db.get(User, p.user_id)
    return _to_out(p, owner.name if owner else "")


@router.patch("/{pid}", response_model=PortfolioOut)
def update_portfolio(
    pid: int,
    payload: PortfolioUpdate,
    user: User = Depends(get_approved_user),
    db: Session = Depends(get_db),
):
    p = db.get(Portfolio, pid)
    if not p:
        raise HTTPException(status_code=404, detail="Not found")
    if p.user_id != user.id and user.role != UserRole.ADMIN.value:
        raise HTTPException(status_code=403, detail="Forbidden")
    if payload.name is not None:
        p.name = payload.name
    if payload.notes is not None:
        p.notes = payload.notes
    if payload.is_public is not None:
        p.is_public = payload.is_public
    db.commit()
    db.refresh(p)
    owner = db.get(User, p.user_id)
    return _to_out(p, owner.name if owner else "")


@router.delete("/{pid}")
def delete_portfolio(pid: int, user: User = Depends(get_approved_user), db: Session = Depends(get_db)):
    p = db.get(Portfolio, pid)
    if not p:
        raise HTTPException(status_code=404, detail="Not found")
    if p.user_id != user.id and user.role != UserRole.ADMIN.value:
        raise HTTPException(status_code=403, detail="Forbidden")
    db.delete(p)
    db.commit()
    return {"ok": True}

File: frontend/src/api/portfolios.ts
import { api } from "./client";

export interface AssetWeight {
  symbol: string;
  name: string;
  category: string;
  weight: number;
  amount_usd: number;
  expected_return_annual: number;
  volatility_annual: number;
  /** Geometric annual return (CAGR) over the asset's own full history.
   *  Null if not provided by backend (e.g. libertex_lite mode). */
  cagr_annual?: number | null;
}

export interface MonteCarloResult {
  n_simulations: number;
  n_months: number;
  initial_capital: number;
  expected_value: number;
  expected_return_pct: number;
  percentiles: { p5: number; p25: number; p50: number; p75: number; p95: number };
  var_95: number;
  cvar_95: number;
  median_path: number[];
  p5_path: number[];
  p25_path: number[];
  p75_path: number[];
  p95_path: number[];
  paths_sample: number[][];
  months: number[];
}

export interface OptimizeRequest {
  portfolio_type: "min_variance" | "max_sharpe" | "target_return" | "target_risk";
  initial_capital: number;
  risk_tolerance: "conservative" | "moderate" | "aggressive";
  target_return?: number | null;
  target_risk?: number | null;
  history_years: number;
  min_history_years: number;
  cov_method: "ledoit_wolf" | "sample" | "ewma";
  long_only: boolean;
  sparsify: boolean;
  sparsify_threshold: number;
  /** Per-asset hard cap on weight. Prevents single-asset "portfolios" at
   *  the feasibility boundary. 0.35 default. Set to 1.0 to disable. */
  max_weight_per_asset?: number;
  /** Subtract Libertex overnight-swap costs from each asset's returns before
   *  optimising. Mirrors what the user actually keeps after holding CFDs on
   *  a Libertex account. Default false — historical-only as a baseline. */
  apply_swaps?: boolean;
  max_assets_in_universe: number;
  categories?: string[] | null;
  /** Tickers to PULL OUT of the universe before optimisation. Matched
   *  case-insensitively against `Asset.symbol`. Backend default is []. */
  exclude_symbols?: string[];
  /** "Drop from peak" filter — exclude assets currently more than X
   *  fraction below their historical peak. 0.60 = drop if last/peak < 0.40.
   *  Set to 1.0 to disable. Backend default 0.60 catches names like
   *  ENPH-2024 (last ~$100 vs peak $329 = 68% below ATH). */
  max_drop_from_peak_pct?: number;
  /** Restrict universe to MetaTrader-tradeable instruments only. After
   *  dedup ~80 assets (mostly blue-chip stocks + popular crypto + a few
   *  FX crosses). Default false = full Libertex CFD pool. */
  mt_only?: boolean;
}

export interface OptimizeResponse {
  portfolio_type: string;
  initial_capital: number;
  weights: AssetWeight[];
  universe_size: number;
  sparsified: boolean;
  expected_return_annual: number;
  /** Geometric mean (CAGR) of the realised portfolio. Compare with arithmetic
   *  `expected_return_annual` to see variance drag — large gap = volatile mix. */
  cagr_annual: number;
  volatility_annual: number;
  sharpe_ratio: number;
  sortino_ratio: number;
  var_95_annual: number;
  cvar_95_annual: number;
  max_drawdown_estimate: number;
  risk_free_rate: number;
  efficient_frontier: Array<Record<string, number>>;
  monte_carlo: MonteCarloResult;
  correlation_matrix: { symbols: string[]; matrix: number[][] };
  benchmark_comparison: any;
  history_years: number;
  min_history_years: number;
  cov_method: string;
  estimation_window_start: string;
  estimation_window_end: string;
  /** Per-run pipeline trace id. Empty string if persistence failed. */
  trace_id?: string;
}

export async function downloadTrace(traceId: string): Promise<void> {
  const r = await api.get(`/optimize/trace/${encodeURIComponent(traceId)}`, {
    responseType: "blob",
  });
  const blob = new Blob([r.data], { type: "text/markdown; charset=utf-8" });
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = `portfolio-trace-${traceId.slice(0, 8)}.md`;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}

export interface PortfolioListItem {
  id: number;
  user_id: number;
  owner_name: string;
  name: string;
  portfolio_type: string;
  initial_capital: number;
  expected_return_annual: number;
  volatility_annual: number;
  sharpe_ratio: number;
  is_public: boolean;
  is_mine: boolean;
  created_at: string;
}

export interface PortfolioOut extends OptimizeResponse {
  id: number;
  user_id: number;
  owner_name: string;
  name: string;
  risk_tolerance: string;
  target_return: number | null;
  target_risk: number | null;
  is_public: boolean;
  notes: string | null;
  created_at: string;
}

export async function optimize(req: OptimizeRequest): Promise<OptimizeResponse> {
  const r = await api.post("/optimize", req);
  return r.data;
}

export async function quotaStatus() {
  const r = await api.get("/optimize/quota-status");
  return r.data;
}

export async function savePortfolio(payload: {
  name: string;
  notes?: string | null;
  is_public: boolean;
  optimize_request: OptimizeRequest;
  optimize_result: OptimizeResponse;
}): Promise<PortfolioOut> {
  const r = await api.post("/portfolios", payload);
  return r.data;
}

export async function listPortfolios(): Promise<{ portfolios: PortfolioListItem[]; total: number }> {
  const r = await api.get("/portfolios");
  return r.data;
}

export async function getPortfolio(id: number): Promise<PortfolioOut> {
  const r = await api.get(`/portfolios/${id}`);
  return r.data;
}

export async function updatePortfolio(id: number, patch: { name?: string; notes?: string; is_public?: boolean }) {
  const r = await api.patch(`/portfolios/${id}`, patch);
  return r.data;
}

export async function deletePortfolio(id: number) {
  await api.delete(`/portfolios/${id}`);
}

export async function downloadFile(id: number, kind: "pdf" | "excel") {
  const r = await api.get(`/export/portfolio/${id}/${kind}`, { responseType: "blob" });
  const blob = new Blob([r.data], {
    type: kind === "pdf" ? "application/pdf" : "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
  });
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = `portfolio_${id}.${kind === "pdf" ? "pdf" : "xlsx"}`;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}
